src/main/scala/wdl4s/WdlNamespace.scala

```diff
@@ -197,85 +167,297 @@ object WdlNamespace {
+    } yield ast
+
+    val children = tasks ++ namespaces ++ nonTaskScopes.map(ast => getScope(ast, parent = None))
+
+    val namespace = children.collect({ case w: Workflow => w }) match {
+      case Nil => WdlNamespaceWithoutWorkflow(namespaceName, imports, namespaces, tasks, terminalMap, ast)
+      case Seq(workflow) => WdlNamespaceWithWorkflow(ast, workflow, namespaceName, imports, namespaces, tasks, terminalMap, wdlSyntaxErrorFormatter)
+      case _ => throw new SyntaxError(wdlSyntaxErrorFormatter.tooManyWorkflows(ast.findAsts(AstNodeName.Workflow).asJava))
     }
 
     /**
-     * All `task` definitions defined in the WDL file (i.e. not imported)
-     */
-    val localTasks: Seq[Task] = ast.findAsts(AstNodeName.Task) map {Task(_, wdlSyntaxErrorFormatter)}
+      * Write-once var setting for parent/child relationships
+      */
+
+    def descendants(scope: Scope): Seq[Scope] = {
+      val children = scope.children
+      val childDescendants = scope.children.flatMap({
+        case n: WdlNamespace => Seq.empty
+        case s => descendants(s)
+      })
+      children ++ childDescendants
+    }
+
+    namespace.children = children
+    namespace.children.foreach(_.parent = namespace)
+
+    tasks foreach { task =>
+      task.children = getChildren(task.ast, Option(task))
+      task.children.foreach(_.parent = task)
+    }
+
+    descendants(namespace).foreach(_.namespace = namespace)
 
     /**
-     * All `task` definitions, including local and imported ones
-     */
-    val tasks: Seq[Task] = localTasks ++ importedTasks
-
-    /* 
-     * Ensure that no namespaces collide with task names. 
-     * 
-     * It'd be simpler to get this via the `namespaces` themselves but don't have access to the correct AST, which is
-     * required by the error syntax highlighter :/ (FIXME: Or do I?)
-     */
-    for {
-      i <- imports
-      namespaceAst <- i.namespaceAst
-      task <- findTask(namespaceAst.sourceString, namespaces, tasks)
-    } yield {throw new SyntaxError(wdlSyntaxErrorFormatter.taskAndNamespaceHaveSameName(task.ast, namespaceAst.asInstanceOf[Terminal]))}
-
-    // Detect duplicated task names
-    val dupeTaskAstsByName = tasks.map(_.ast).duplicatesByName
-    if (dupeTaskAstsByName.nonEmpty) {
-      throw new SyntaxError(wdlSyntaxErrorFormatter.duplicateTask(dupeTaskAstsByName))
+      * SYNTAX CHECKS
+      */
+
+    val callInputSectionErrors = namespace.descendants.collect({ case c: Call => c }).flatMap(
+      validateCallInputSection(_, wdlSyntaxErrorFormatter)
+    )
+
+    val declarationErrors = for {
+      descendant <- namespace.descendants
+      declaration <- getDecls(descendant)
+      error <- validateDeclaration(declaration, wdlSyntaxErrorFormatter)
+    } yield error
+
+    def scopeNameAndTerminal(scope: Scope): (String, Terminal) = {
+      scope match {
+        case ns: WdlNamespace => ("Namespace", imports.find(_.namespaceName == ns.importedAs).get.namespaceTerminal.get)
```

Comment:
given that it's called only when there are duplicate sibling scopes, I would say it's safe because if we got there it means that there has to be an imported namespace that shares its name with the current. But I'll see if I can make this more `None.get` safe